ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java

```diff
@@ -0,0 +1,350 @@
+              this::getTrieNode, worldStateRootHash, Function.identity(), Function.identity());
+
+      // for manicured tries and composting, collect branches here (not implemented)
+
+      // now add the accounts
+      for (final Map.Entry<Address, BonsaiValue<BonsaiAccount>> accountUpdate :
+          updater.getAccountsToUpdate().entrySet()) {
+        final Bytes accountKey = accountUpdate.getKey();
+        final BonsaiValue<BonsaiAccount> bonsaiValue = accountUpdate.getValue();
+        final BonsaiAccount updatedAccount = bonsaiValue.getUpdated();
+        if (updatedAccount == null) {
+          final Hash addressHash = Hash.hash(accountKey);
+          accountTrie.remove(addressHash);
+          accountTx.remove(accountKey.toArrayUnsafe());
+        } else {
+          final Hash addressHash = updatedAccount.getAddressHash();
+          final Bytes accountValue = updatedAccount.serializeAccount();
+          accountTx.put(accountKey.toArrayUnsafe(), accountValue.toArrayUnsafe());
+          accountTrie.put(addressHash, accountValue);
+        }
+      }
+
+      accountTrie.commit(
+          (location, hash, value) -> writeTrieNode(trieBranchTx, location, hash, value));
+      worldStateRootHash = accountTrie.getRootHash();
+      trieBranchTx.put(WORLD_ROOT_KEY, worldStateRootHash.toArrayUnsafe());
+
+      // for manicured tries and composting, trim and compost branches here
+
+      if (blockHash != null) {
+        final TrieLogLayer trieLog = updater.generateTrieLog(blockHash);
+        trieLog.freeze();
+        // TODO add to archive here, but only once we get persisted follow distance implemented
+        // archive.addLayeredWorldState(new BonsaiLayeredWorldState(this, trieLog));
+
+        final BytesValueRLPOutput rlpLog = new BytesValueRLPOutput();
+        trieLog.writeTo(rlpLog);
+        trieLogTx.put(blockHash.toArrayUnsafe(), rlpLog.encoded().toArrayUnsafe());
+      }
+
+      success = true;
+    } finally {
+      if (success) {
+        accountTx.commit();
+        codeTx.commit();
+        storageTx.commit();
+        trieBranchTx.commit();
+        trieLogTx.commit();
+        updater.reset();
+      } else {
+        accountTx.rollback();
+        codeTx.rollback();
+        storageTx.rollback();
+        trieBranchTx.rollback();
+        trieLogTx.rollback();
+      }
+    }
+  }
+
+  private static Bytes rlpEncode(final Bytes bytes) {
+    final BytesValueRLPOutput out = new BytesValueRLPOutput();
+    out.writeBytes(bytes.trimLeadingZeros());
+    return out.encoded();
+  }
+
+  @Override
+  public WorldUpdater updater() {
+    if (updater == null) {
+      updater = new BonsaiWorldStateUpdater(this);
+    }
+    return updater;
+  }
+
+  @Override
+  public Hash rootHash() {
+    return Hash.wrap(worldStateRootHash);
+  }
+
+  @Override
+  public Stream<StreamableAccount> streamAccounts(final Bytes32 startKeyHash, final int limit) {
+    throw new RuntimeException("NIY");
```

Comment:
NIY ? Not implemented yet ?

Reply:
Not Implemented Yet.  Yes.  I'll put a more verbose description.

Reply:
The impact is the debug API and hence remix needs it.  So I would have to do a heavy weight merge operation with the layered modes.  For local debugging I think keeping forest for now is the thing to do.